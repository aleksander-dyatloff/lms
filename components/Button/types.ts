import { ColorPalette } from '@styles/theme/palette'

namespace ButtonComponent {
  export interface WrapperProps {
    palette: ColorPalette
  }

  export interface Props {
    palette?: ColorPalette
    className?: string
    variant?: 'filled' | 'outlined' | 'text'
  }
}

export default ButtonComponent
