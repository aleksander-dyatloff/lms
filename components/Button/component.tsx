import combineClasses from '@utils/combineClasses'
import { FC } from 'react'

import Wrapper from './styles'
import ButtonComponent from './types'

const Button: FC<ButtonComponent.Props> = ({
  variant = 'filled',
  className,
  ...restProps
}) => {
  const wrapperClasses = combineClasses(variant, className)

  return (
    <Wrapper
      as='button'
      className={wrapperClasses}
      {...restProps}
    />
  )
}

export default Button
