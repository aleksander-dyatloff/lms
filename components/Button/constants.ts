import ButtonComponent from './types'

export const variantClasses = {
  [ButtonComponent.Variant.Filled]: 'filled',
  [ButtonComponent.Variant.Outlined]: 'outlined',
  [ButtonComponent.Variant.Text]: 'text',
}
