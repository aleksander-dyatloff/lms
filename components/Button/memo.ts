import { memo } from 'react'

import Button from './component'

export default memo(Button)
