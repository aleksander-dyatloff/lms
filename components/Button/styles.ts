import Element from '@components/Element/styles'
import styled from 'styled-components'

import ButtonComponent from './types'

const Wrapper = styled(Element)<ButtonComponent.WrapperProps>`
  --mainColor: ${({
    theme, palette,
  }) => palette?.main ?? theme.palette.primary.main};
  --contrastColor: ${({
    theme, palette,
  }) => palette?.contrast ?? theme.palette.primary.contrast};

  display: flex;
  align-items: center;
  justify-content: center;
  min-height: 36px;
  cursor: pointer;
  border-radius: ${({ theme }) => theme.borderRadius * 3}px;
  padding: ${({ theme }) => `${theme.spacing}px ${theme.spacing * 3}`}px;
  transition-property: background-color, border-color, color;
  transition-duration: 160ms;

  &:not([disabled]):hover {
    --mainColor: ${({
    theme, palette,
  }) => palette?.mainDarken ?? theme.palette.primary.mainDarken};
  }

  &.filled {
    background-color: var(--mainColor);
    color: var(--contrastColor);
  }

  &.outlined {
    color: var(--mainColor);
    border: 2px solid var(--mainColor);
  }

  &.text {
    color: var(--mainColor);  
  }
`

export default Wrapper
