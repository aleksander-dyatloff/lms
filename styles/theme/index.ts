import { DefaultTheme } from 'styled-components'

import { palette } from './palette'

const theme: DefaultTheme = {
  spacing: 4,
  borderRadius: 4,
  palette,
}

export type ITheme = typeof theme

export default theme
