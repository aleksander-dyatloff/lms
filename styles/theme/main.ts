import { DefaultTheme } from 'styled-components'

const mainTheme: DefaultTheme = {
  spacing: 4,
  borderRadius: 4,
  palette: {
    primary: {
      main: '#8e24aa',
      mainDarken: '#6d1785',
      contrast: '#ffffff',
      error: 'red',
    },
    secondary: {
      main: '#ffffff',
      mainDarken: '#f2f2f2',
      contrast: '#424242',
      error: 'red',
    },
  },
}

export type ITheme = typeof mainTheme

export default mainTheme
