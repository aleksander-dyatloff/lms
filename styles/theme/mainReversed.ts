import { DefaultTheme } from 'styled-components'

const mainReversedTheme: DefaultTheme = {
  spacing: 4,
  borderRadius: 4,
  palette: {
    primary: {
      main: '#ffffff',
      mainDarken: '#f2f2f2',
      contrast: '#8e24aa',
      error: 'red',
    },
    secondary: {
      main: '#ffffff',
      mainDarken: '#f2f2f2',
      contrast: '#424242',
      error: 'red',
    },
  },
}

export type ITheme = typeof mainReversedTheme

export default mainReversedTheme
