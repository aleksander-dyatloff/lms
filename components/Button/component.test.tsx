import 'jest-styled-components'

import theme from '@styles/theme'
import { render } from '@testing-library/react'
import { Profiler } from 'react'
import { ThemeProvider } from 'styled-components'

import Button from './component'

describe('Button component', () => {
  test('should match to snapshot', () => {
    const { container } = render(
      <ThemeProvider theme={theme}>
        <Button>
          Click
        </Button>
      </ThemeProvider>,
    )

    expect(container).toMatchSnapshot()
  })

  test('first render duration should match to specification', () => {
    render(
      <ThemeProvider theme={theme}>
        <Profiler
          id='Button'
          onRender={(_, __, renderDuration) => {
            expect(renderDuration).toBeLessThanOrEqual(5)
          }}
        >
          <Button>
            Click
          </Button>
        </Profiler>
      </ThemeProvider>,
    )
  })
})
