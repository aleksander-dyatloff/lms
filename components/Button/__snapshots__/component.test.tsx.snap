// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button component should match to snapshot 1`] = `
.c0 {
  padding: 0;
  margin: 0;
  font: inherit;
  -webkit-letter-spacing: inherit;
  -moz-letter-spacing: inherit;
  -ms-letter-spacing: inherit;
  letter-spacing: inherit;
  list-style: none;
  color: inherit;
  background-color: inherit;
  border: none;
  border-radius: 0;
  width: 100%;
  text-align: inherit;
  -webkit-text-decoration: inherit;
  text-decoration: inherit;
  cursor: inherit;
  outline: none;
  caret-color: inherit;
}

.c1 {
  --mainColor: #8e24aa;
  --contrastColor: #ffffff;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  min-height: 36px;
  cursor: pointer;
  border-radius: 12px;
  padding: 4px 12px;
  -webkit-transition-property: background-color,border-color,color;
  transition-property: background-color,border-color,color;
  -webkit-transition-duration: 160ms;
  transition-duration: 160ms;
}

.c1:not([disabled]):hover {
  --mainColor: #6d1785;
}

.c1.filled {
  background-color: var(--mainColor);
  color: var(--contrastColor);
}

.c1.outlined {
  color: var(--mainColor);
  border: 2px solid var(--mainColor);
}

.c1.text {
  color: var(--mainColor);
}

<div>
  <button
    class="c0 c1  filled"
  >
    Click
  </button>
</div>
`;
